torchaudio/models/_tacotron2.py

```diff
@@ -0,0 +1,1048 @@
+            mask: binary mask for padded data
+        """
+
+        alignment = self._get_alignment_energies(
+            attention_hidden_state, processed_memory, attention_weights_cat
+        )
+        alignment = alignment.masked_fill(mask, self.score_mask_value)
+        attention_weights = F.softmax(alignment, dim=1)
+
+        attention_context = torch.bmm(attention_weights.unsqueeze(1), memory)
+        attention_context = attention_context.squeeze(1)
+
+        return attention_context, attention_weights
+
+
+class _Encoder(nn.Module):
+    r"""Encoder Module
+
+        Args:
+            n_encoder_convolutions: the number of convolution layers in the encoder.
+            n_encoder_embedding: the number of embedding dimensions in the encoder.
+            n_encoder_kernel_size: the kernel size in the encoder.
+
+        Examples
+            >>> encoder = _Encoder(3, 512, 5)
+            >>> input = torch.rand(10, 20, 30)
+            >>> output = encoder(input)  # shape: (10, 30, 512)
+    """
+
+    def __init__(
+        self, n_encoder_convolutions, n_encoder_embedding, n_encoder_kernel_size
+    ) -> None:
+        super(_Encoder, self).__init__()
+
+        convolutions = []
+        for _ in range(n_encoder_convolutions):
+            conv_layer = nn.Sequential(
+                _ConvNorm(
+                    n_encoder_embedding,
+                    n_encoder_embedding,
+                    kernel_size=n_encoder_kernel_size,
+                    stride=1,
+                    padding=int((n_encoder_kernel_size - 1) / 2),
+                    dilation=1,
+                    w_init_gain="relu",
+                ),
+                nn.BatchNorm1d(n_encoder_embedding),
+            )
+            convolutions.append(conv_layer)
+
+        self.convolutions = nn.ModuleList(convolutions)
+
+        self.lstm = nn.LSTM(
+            n_encoder_embedding,
+            int(n_encoder_embedding / 2),
+            1,
+            batch_first=True,
+            bidirectional=True,
+        )
+
+    def forward(self, x: Tensor, input_lengths: Tensor) -> Tensor:
+        r"""Pass the input through the _Encoder layer.
+
+        Args:
+            x (Tensor): the input sequence to the _Encoder layer (n_batch, n_encoder_embedding, n_seq).
+            input_lengths (Tensor): the length of input sequence to the _Encoder layer (n_batch,).
+
+        Return:
+            Tensor shape: (n_batch, n_seq, n_encoder_embedding)
+        """
+
+        for conv in self.convolutions:
+            x = F.dropout(F.relu(conv(x)), 0.5, self.training)
+
+        x = x.transpose(1, 2)
+
+        x = nn.utils.rnn.pack_padded_sequence(x, input_lengths, batch_first=True)
+
+        self.lstm.flatten_parameters()
+        outputs, _ = self.lstm(x)
+        outputs, _ = nn.utils.rnn.pad_packed_sequence(outputs, batch_first=True)
+
+        return outputs
+
+    def infer(self, x: Tensor, input_lengths: Tensor) -> Tensor:
```

Comment:
What's the difference with forward here?

Reply:
The difference here is the dropout layer. I agree, since the "infer" won't be used in the training, I will remove this part.